### PR TITLE
Dockerfile Optimization

### PR DESCRIPTION
**Pull request by dmccoystephenson**
## Changes
Some changes have been made to the Dockerfiles with the aim to enhance performance and efficiency:

1. Base Image: The base image in the Dockerfiles has been switched from Ubuntu to Alpine, resulting in smaller and faster deployments.
1. Splitting Dockerfiles: The Dockerfiles are now divided into two images - the builder image and the runtime image. The project is built within the builder container, and only the necessary files are copied to a new container for deployment.

Additionally, some output from a cv-lib method has been decreased in precision to fix a failing unit test in alpine.

## Testing
The changes have been thoroughly tested with the following results:

1. Successful Compilation: The project compiles and starts up successfully.
1. Passing Unit Tests: All unit tests pass within the Alpine container.
1. Dev Cluster Deployment: When deployed to the development cluster, the data gets processed as expected.

These updates are expected to contribute to a more efficient and streamlined development and deployment process.

### CODE REVIEW

File: Dockerfile-nsv [hunk 1]
```diff
@@ -34,11 +30,27 @@ ADD config/${PPM_CONFIG_FILE} /cvdi-stream/config/
...
 
-# Add test data. This changes frequently so keep it low in the file.
+# === RUNTIME IMAGE ===
+FROM alpine:3.12
+USER root
+
+WORKDIR /cvdi-stream
+
+# add runtime dependencies
+RUN apk add --upgrade --no-cache \
```
Comment (dan-du-car) [inline Dockerfile-nsv:43]: Is it possible to COPY the libs from build image?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -43,11 +43,11 @@ jobs:
          cmake --build _cmake_build
          cmake --build _cmake_build --target install  
      - name: Install sonar-scanner and build-wrapper
-       uses: SonarSource/sonarcloud-github-c-cpp@v1  # This Action Installs sonar cloud and build wrapper to run sonar scan analysis  
+       uses: sonarsource/sonarcloud-github-c-cpp@v2  # This Action Installs sonar cloud and build wrapper to run sonar scan analysis  
      - name: Set up Docker Buildx
-       uses: docker/setup-buildx-action@v2
```
Comment (dan-du-car) [inline .github/workflows/ci.yml:48]: Remove the docker setup action

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -43,11 +43,11 @@ jobs:
          cmake --build _cmake_build
          cmake --build _cmake_build --target install  
      - name: Install sonar-scanner and build-wrapper
-       uses: SonarSource/sonarcloud-github-c-cpp@v1  # This Action Installs sonar cloud and build wrapper to run sonar scan analysis  
+       uses: sonarsource/sonarcloud-github-c-cpp@v2  # This Action Installs sonar cloud and build wrapper to run sonar scan analysis  
      - name: Set up Docker Buildx
-       uses: docker/setup-buildx-action@v2
+       uses: docker/setup-buildx-action@v3
      - name: Build
-       uses: docker/build-push-action@v3
```
Comment (dan-du-car) [inline .github/workflows/ci.yml:50]: remove docker build/push action

File: Dockerfile-nsv (hunk 1 above)
Comment (dmccoystephenson) [reply Dockerfile-nsv]: The dockerfiles have now been modified to include 3 image stages:
1. runtime-deps: Image with runtime dependencies
2. builder: Inherit from runtime-deps and build
3. final: Inherit from runtime-deps and copy built files

This removes the need to install the runtime dependencies twice.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (dmccoystephenson) [reply .github/workflows/ci.yml:48]: This has been removed.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (dmccoystephenson) [reply .github/workflows/ci.yml:50]: This has been removed.

**Comment by dmccoystephenson** [comment]
This PR is ready for re-review.

**Review by dan-du-car — APPROVED** [review]

**Pull request merged**